mmdet/models/dense_heads/corner_head.py

```diff
@@ -0,0 +1,958 @@
+from math import ceil, log
+
+import torch
+import torch.nn as nn
+from mmcv.cnn import ConvModule, bias_init_with_prob
+
+from mmdet.core import multi_apply
+from mmdet.ops import CornerPool, batched_nms
+from ..builder import HEADS, build_loss
+from ..utils import gaussian_radius, gen_gaussian_target
+
+
+class BCPool(nn.Module):
+    """Bidirectional Corner Pooling Module (TopLeft, BottomRight, etc.)
+
+    Args:
+        in_channels (int): Input channels of module.
+        pool_direction1 (str): direction of the first CornerPool.
+        pool_direction2 (str): direction of the second CornerPool.
+        norm_cfg (dict): Dictionary to construct and config norm layer.
+    """
+
+    def __init__(self,
+                 in_channels,
+                 pool_direction1,
+                 pool_direction2,
+                 feat_channels=128,
+                 norm_cfg=dict(type='BN', requires_grad=True)):
+        super(BCPool, self).__init__()
+        self.pool1_conv = ConvModule(
+            in_channels, feat_channels, 3, padding=1, norm_cfg=norm_cfg)
+        self.pool2_conv = ConvModule(
+            in_channels, feat_channels, 3, padding=1, norm_cfg=norm_cfg)
+
+        self.p_conv = ConvModule(
+            feat_channels,
+            in_channels,
+            3,
+            padding=1,
+            norm_cfg=norm_cfg,
+            act_cfg=None)
+
+        self.conv1 = ConvModule(
+            in_channels, in_channels, 1, norm_cfg=norm_cfg, act_cfg=None)
+        self.conv2 = ConvModule(
+            in_channels, in_channels, 3, padding=1, norm_cfg=norm_cfg)
+
+        self.pool1 = CornerPool(pool_direction1)
+        self.pool2 = CornerPool(pool_direction2)
+        self.relu = nn.ReLU(inplace=True)
+
+    def forward(self, x):
+        """Forward features from the upstream network.
+
+        Args:
+            x (tensor): Input feature of BCPool.
+
+        Returns:
+            conv2 (tensor): Output feature of BCPool.
+        """
+        pool1_conv = self.pool1_conv(x)
+        pool2_conv = self.pool2_conv(x)
+        pool1_feat = self.pool1(pool1_conv)
+        pool2_feat = self.pool2(pool2_conv)
+        p_conv = self.p_conv(pool1_feat + pool2_feat)
+        conv1 = self.conv1(x)
+        relu = self.relu(p_conv + conv1)
+        conv2 = self.conv2(relu)
+        return conv2
+
+
+@HEADS.register_module()
+class CornerHead(nn.Module):
```

Comment:
How about the speed compared to the official repo?

Reply:
On 1080TI with 3 images per gpu, our speed is about 1.62s/iter, the official repo is 1.79s/iter.
On V100 + torch1.5 and using `torch.cummax` instead of corner_pool could also get ~10% speedup. #3008